src/Compilers/CSharp/Test/Semantic/Diagnostics/DiagnosticAnalyzerTests.cs

```diff
@@ -872,23 +884,59 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
             public override void Initialize(AnalysisContext context)
             {
                 context.RegisterCompilationAction(compilationContext =>
-                    compilationContext.ReportDiagnostic(CodeAnalysis.Diagnostic.Create(UnsupportedDescriptor, Location.None)));
+                {
+                    try
+                    {
+                        ThrownException = null;
+                        compilationContext.ReportDiagnostic(CodeAnalysis.Diagnostic.Create(UnsupportedDescriptor, Location.None));
+                    }
+                    catch (Exception e)
+                    {
+                        ThrownException = e;
+                        throw;
+                    }
+                });
             }
         }
 
         [Fact, WorkItem(4376, "https://github.com/dotnet/roslyn/issues/4376")]
         public void TestReportingDiagnosticWithInvalidId()
         {
             string source = @"";
+            CSharpCompilation compilation = CreateCompilationWithMscorlib45(source);
             var analyzers = new DiagnosticAnalyzer[] { new AnalyzerWithInvalidDiagnosticId() };
             string message = new ArgumentException(string.Format(CodeAnalysisResources.InvalidDiagnosticIdReported, AnalyzerWithInvalidDiagnosticId.Descriptor.Id), "diagnostic").Message;
+            Exception analyzerException = null;
+            IFormattable context = $@"{string.Format(CodeAnalysisResources.ExceptionContext, $@"Compilation: {compilation.AssemblyName}")}
 
-            CreateCompilationWithMscorlib45(source)
-                .VerifyDiagnostics()
-                .VerifyAnalyzerDiagnostics(analyzers, null, null, logAnalyzerExceptionAsDiagnostics: true,
-                     expected: Diagnostic("AD0001")
-                     .WithArguments("Microsoft.CodeAnalysis.CommonDiagnosticAnalyzers+AnalyzerWithInvalidDiagnosticId", "System.ArgumentException", message)
-                     .WithLocation(1, 1));
+{new LazyToString(() => analyzerException)}
+-----
+
+{string.Format(CodeAnalysisResources.DisableAnalyzerDiagnosticsMessage, "Invalid ID")}";
+
+            EventHandler<FirstChanceExceptionEventArgs> firstChanceException = (sender, e) =>
+            {
+                if (e.Exception is ArgumentException)
+                {
+                    analyzerException = e.Exception;
+                }
+            };
+
+            try
+            {
+                AppDomain.CurrentDomain.FirstChanceException += firstChanceException;
```

Comment:
> AppDomain.CurrentDomain.FirstChanceException += firstChanceException; [](start = 15, length = 70)

This is really brittle to do in our unit test environment where there may be stray `Task` running from a previous test run. If we need to actually catch first chance exceptions we should just do this in a separate `AppDomain`

Reply:
I believe we lose the ability to run the tests on other platforms in that case. Note that we would only get interference if an `ArgumentException` was thrown from another test, and if the test did fail it would include a stack trace for the code running from a previous test.

Reply:
Okay then we need another approach to validating this change. As written the validation is flaky and can't be safely merged.

Reply:
What about `e.Exception is ArgumentException { Message: message }`?

Reply:
@jaredpar would the proposed change address your concerns?

Reply:
@jaredpar If your review is blocking, can you mark this as Request Changes? Also, I would like to get this merged for 16.5 Preview 2 but I need to know if the above proposed test strategy is acceptable.